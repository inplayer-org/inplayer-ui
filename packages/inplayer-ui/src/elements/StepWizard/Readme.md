A software wizard or setup assistant is a user interface type that presents a user with a sequence of dialog boxes that lead the user through a series of well-defined steps. Tasks that are complex, infrequently performed, or unfamiliar may be easier to perform using a wizard.

```jsx static
import React from 'react';
import { StepWizard } from '@inplayer-org/inplayer-ui';

function StepOne() {
  return (
    <div>
      <Typography variant="h2">Step One!</Typography>
    </div>
  );
};

function StepTwo() {
  return (
    <div>
      <Typography variant="h2">Step Two!</Typography>
    </div>
  );
};

function StepThree() {
  return (
    <div>
      <Typography variant="h2">Step Three!</Typography>
    </div>
  );
};

function StepFour() {
  return (
    <div>
      <Typography variant="h2">Step Four!</Typography>
    </div>
  );
};

const steps = [
  { isDisabled: false, isCompleted: true, component: <StepOne/> },
  { isDisabled: false, isCompleted: false, component: <StepTwo/> },
  { isDisabled: true, isCompleted: false, component: <StepThree/> },
  { isDisabled: true, isCompleted: false, component: <StepFour/> },
];

class StepWizardExample extends React.Component {
  constructor() {
    super();

    this.state = {
      activeStep: 0,
    };

    this.onStepChange = this.onStepChange.bind(this);
  }

  onStepChange(activeStep) {
    this.setState({ activeStep });
  }

  render() {
    const { activeStep } = this.state;

    return (
      <StepWizard activeStep={activeStep} onStepChange={onStepChange} steps={steps} />
    );
  }
}
```

```jsx
function StepOne() {
  return (
    <div>
      <Typography variant="h2">Step One!</Typography>
    </div>
  );
};

function StepTwo() {
  return (
    <div>
      <Typography variant="h2">Step Two!</Typography>
    </div>
  );
};

function StepThree() {
  return (
    <div>
      <Typography variant="h2">Step Three!</Typography>
    </div>
  );
};

function StepFour() {
  return (
    <div>
      <Typography variant="h2">Step Four!</Typography>
    </div>
  );
};

const steps = [
  { isDisabled: false, isCompleted: true, component: <StepOne/> },
  { isDisabled: false, isCompleted: false, component: <StepTwo/> },
  { isDisabled: true, isCompleted: false, component: <StepThree/> },
  { isDisabled: true, isCompleted: false, component: <StepFour/> },
];


class StepWizardExample extends React.Component {
  constructor() {
    super();

    this.state = {
      activeStep: 0,
    };

    this.onStepChange = this.onStepChange.bind(this);
  }

   onStepChange(activeStep) {
    this.setState({ activeStep });
  }

  render() {
    const { activeStep } = this.state;

    return (
       <StepWizard activeStep={activeStep} onStepChange={this.onStepChange} steps={steps} />
    );
  }
}
<StepWizardExample />;
```
