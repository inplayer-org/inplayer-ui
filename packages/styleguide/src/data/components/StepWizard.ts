import { ComponentData } from "../types";

const usage = `
  import React from 'react';
  import { StepWizard } from '@inplayer-org/inplayer-ui';

  function StepOne() {
    return (
      <div>
        <Typography variant="h2">Step One!</Typography>
      </div>
    );
  };

  function StepTwo() {
    return (
      <div>
        <Typography variant="h2">Step Two!</Typography>
      </div>
    );
  };

  function StepThree() {
    return (
      <div>
        <Typography variant="h2">Step Three!</Typography>
      </div>
    );
  };

  function StepFour() {
    return (
      <div>
        <Typography variant="h2">Step Four!</Typography>
      </div>
    );
  };

  const steps = [
    { isDisabled: false, isCompleted: true, component: <StepOne/> },
    { isDisabled: false, isCompleted: false, component: <StepTwo/> },
    { isDisabled: true, isCompleted: false, component: <StepThree/> },
    { isDisabled: true, isCompleted: false, component: <StepFour/> },
  ];

  class StepWizardExample extends React.Component {
    constructor() {
      super();

      this.state = {
        activeStep: 0,
      };

      this.onStepChange = this.onStepChange.bind(this);
    }

    onStepChange(activeStep) {
      this.setState({ activeStep });
    }

    render() {
      const { activeStep } = this.state;

      return (
        <StepWizard activeStep={activeStep} onStepChange={onStepChange} steps={steps} />
      );
    }
  }
`

const demoCode = `
  class StepWizardExample extends React.Component {
    constructor() {
      super();

      this.state = {
        activeStep: 0,
        steps: [
          { isDisabled: false, isCompleted: true, 
            component:<Typography variant="h2">Step One!</Typography> },
          { isDisabled: false, isCompleted: false,
            component: <Typography variant="h2">Step Two!</Typography> },
          { isDisabled: true, isCompleted: false, 
            component: <Typography variant="h2">Step Three!</Typography> },
          { isDisabled: true, isCompleted: false,
            component: <Typography variant="h2">Step Four!</Typography> },
        ]
      };

      this.onStepChange = this.onStepChange.bind(this);
    }

    onStepChange(activeStep) {
      this.setState({ activeStep });
    }

    render() {
      const { activeStep } = this.state;

      return (
        <StepWizard activeStep={activeStep} onStepChange={this.onStepChange} steps={this.state.steps} />
      );
    }
  }
`

const StepWizard: ComponentData = {
  description: `A software wizard or setup assistant is a user interface type that presents a user with a sequence of dialog boxes that lead the user through a series of well-defined steps. Tasks that are complex, infrequently performed, or unfamiliar may be easier to perform using a wizard.`,
  path: `src/components/StepWizard/index.js`,
  usage,
  examples: [
    {
      title: 'Demo',
      code: demoCode,
    }
  ],
  propsAndMethods: [
    { name: 'activeStep', type:'number', default: 'Required' },
    { name: 'steps', type:'Array<StepItem>', default: 'Required' },
    { name: 'className', type:'string' },
    { name: 'icon', type:'union' },
    { name: 'iconModifiers', type:'Array<string>' },
    { name: 'iconPosition', type:'string' },
    { name: 'size', type:'enum', default: 'md', description: 'One of: xs, sm, md, lg' },
    { name: 'style', type:'Object', default: '{}' },
  ]
}



export default StepWizard;