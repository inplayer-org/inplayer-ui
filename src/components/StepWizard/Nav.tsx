import React from 'react';
import styled, { css } from 'styled-components';
import { ifProp } from 'styled-tools';

import colors from '../../theme/colors';

const StepNavigation = styled.div`
  margin-bottom: 2rem;
  text-align: center;
`;

const StepAction = styled.span<{
  isActive: boolean;
  isCompleted: boolean;
  isDisabled: boolean;
}>`
  cursor: pointer;
  font-size: ${({ theme }) => theme.font.sizes.extraLarge};
  line-height: 1;
  margin: 0 1.5rem;
  opacity: 0.4;
  text-shadow: none;
  transition: opacity 1s ease, text-shadow 1s ease;
  will-change: opacity, text-shadow;
  ${ifProp(
    'isActive',
    css`
      color: ${colors.skyBlue};
      opacity: 1;
      text-shadow: 0 0px 8px;
    `
  )};
  ${ifProp(
    'isDisabled',
    css`
      pointer-events: none;
    `
  )};
  ${ifProp(
    'isCompleted',
    css`
      color: ${colors.green};
      opacity: 1;
      text-shadow: 0 0px 8px;
    `
  )};
`;

export type StepItem = {
  isDisabled: boolean;
  isCompleted: boolean;
  component: React.ReactChild;
};

type Props = {
  steps: Array<StepItem>;
  activeStep: number;
  goToStep: (activeStep: number) => void;
};

const Nav = ({ steps, goToStep, activeStep }: Props) => (
  <StepNavigation>
    {steps.map(({ isDisabled, isCompleted }, i) => (
      <StepAction
        key={`step-${i}`}
        isActive={i === activeStep}
        isCompleted={isCompleted}
        onClick={() => goToStep(i)}
        isDisabled={isDisabled}
      >
        &bull;
      </StepAction>
    ))}
  </StepNavigation>
);

export default Nav;
