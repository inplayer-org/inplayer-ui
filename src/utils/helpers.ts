import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
