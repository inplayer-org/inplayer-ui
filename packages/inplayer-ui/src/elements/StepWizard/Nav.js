// @flow
import React from 'react';
import styled, { css } from 'styled-components';
import { ifProp } from 'styled-tools';
import { uiColors, fontSizes } from 'utils';
import colors from 'config/colors';

const StepNavigation = styled.div`
  margin-bottom: 2rem;
  text-align: center;
`;

const StepAction = styled.span`
  cursor: pointer;
  font-size: ${fontSizes('extraLarge')};
  line-height: 1;
  margin: 0 1.5rem;
  opacity: 0.4;
  text-shadow: none;
  transition: opacity 1s ease, text-shadow 1s ease;
  will-change: opacity, text-shadow;
  ${ifProp(
    'isActive',
    css`
      color: ${uiColors('primary.main')};
      opacity: 1;
      text-shadow: 0 0px 8px;
    `
  )};
  ${ifProp(
    'isDisabled',
    css`
      pointer-events: none;
    `
  )};
  ${ifProp(
    'isCompleted',
    css`
      color: ${colors.green};
      opacity: 1;
      text-shadow: 0 0px 8px;
    `
  )};
`;

export type StepItem = {
  isDisabled: boolean,
  isCompleted: boolean,
  component: React.ReactNode,
};

type Props = {
  steps: Array<StepItem>,
  activeStep: number,
  goToStep: (activeStep: number) => void,
};

const Nav = ({ steps, goToStep, activeStep }: Props) => (
  <StepNavigation>
    {steps.map(({ isDisabled, isCompleted }, i) => (
      <StepAction
        key={`step-${i}`}
        isActive={i === activeStep}
        isDisabled={isDisabled}
        isCompleted={isCompleted}
        onClick={() => goToStep(i)}
      >
        &bull;
      </StepAction>
    ))}
  </StepNavigation>
);

export default Nav;
