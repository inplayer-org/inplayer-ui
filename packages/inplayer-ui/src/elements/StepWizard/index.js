// @flow
import React from 'react';
import Nav from './Nav';
import Step from './Step';
import type { StepItem } from './Nav';

type TransitionVariant = 'fadeInLeft' | 'fadeInRight' | 'fadeOutLeft' | 'fadeOutRight';

type Props = {
  steps: Array<StepItem>,
  activeStep: number,
  className?: string,
  style?: Object,
  onStepChange?: (activeStep: number) => void,
  transition?: TransitionVariant,
};

const StepWizard = ({ steps, className, style, activeStep, onStepChange, transition }: Props) => {
  const isInvalidStep = currentStep => currentStep < 0 || currentStep > steps.length;

  const setActiveStep = currentStep => {
    if (activeStep === currentStep) return;
    if (isInvalidStep(currentStep)) {
      console.error(`${currentStep} is an invalid step`);
      return;
    }

    // User callback
    onStepChange(currentStep);
  };

  /** Go to step index */
  const goToStep = step => setActiveStep(step);

  const renderSteps = steps.map((step, i) => {
    const isStepActive = i === activeStep;

    if (isStepActive) {
      return (
        <Step key={`step-${i}`} isActive={isStepActive} transition={transition}>
          {step.component}
        </Step>
      );
    }

    return null;
  });

  return (
    <div className={className} style={style}>
      <Nav steps={steps} activeStep={activeStep} goToStep={goToStep} />
      <div>{renderSteps}</div>
    </div>
  );
};

StepWizard.defaultProps = {
  className: null,
  style: {},
  transition: 'fadeInRight',
  onStepChange: () => {},
};

export default StepWizard;
