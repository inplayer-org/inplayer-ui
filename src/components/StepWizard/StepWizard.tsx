import React from 'react';
import Nav, { StepItem } from './Nav';
import Step from './Step';

type TransitionVariant = 'fadeInLeft' | 'fadeInRight' | 'fadeOutLeft' | 'fadeOutRight';

type StepWizardProps = {
  /**
   * Array of objects containing steps information and components
   */
  steps: Array<StepItem>;
  /**
   * Active step of step wizard
   */
  activeStep: number;
  /**
   * Function which runs when the step changes
   */
  onStepChange?: (activeStep: number) => void;
  /**
   * CSS transition type
   */
  transition?: TransitionVariant;
};

const StepWizard = ({
  steps,
  activeStep,
  onStepChange = () => {},
  transition = 'fadeInRight',
}: StepWizardProps) => {
  const isInvalidStep = (currentStep: number) => currentStep < 0 || currentStep > steps.length;

  const setActiveStep = (currentStep: number) => {
    if (activeStep === currentStep) return;
    if (isInvalidStep(currentStep)) {
      console.error(`${currentStep} is an invalid step`);
      return;
    }

    // User callback
    onStepChange(currentStep);
  };

  /** Go to step index */
  const goToStep = (step: number) => setActiveStep(step);

  const renderSteps = steps.map((step, i) => {
    const isStepActive = i === activeStep;

    if (isStepActive) {
      return (
        <Step key={`step-${i}`} isActive={isStepActive} transition={transition}>
          {step.component}
        </Step>
      );
    }

    return null;
  });

  return (
    <>
      <Nav steps={steps} activeStep={activeStep} goToStep={goToStep} />
      <div>{renderSteps}</div>
    </>
  );
};

export default StepWizard;
