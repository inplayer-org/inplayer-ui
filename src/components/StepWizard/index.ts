import StepWizard from './StepWizard';

export default StepWizard;
