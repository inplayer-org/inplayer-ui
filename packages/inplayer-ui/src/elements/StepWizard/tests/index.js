import React from 'packages/inplayer-ui/src/elements/StepWizard/tests/react';
import { generateSnapshots } from 'packages/inplayer-ui/src/elements/StepWizard/tests/helpers';

import StepWizard from '../index';

const StepOne = () => (
  <div>
    <h1 variant="h2">Step One!</h1>
  </div>
);

const StepTwo = () => (
  <div>
    <h1 variant="h2">Step Two!</h1>
  </div>
);

const StepThree = () => (
  <div>
    <h1 variant="h2">Step Three!</h1>
  </div>
);

const steps = [
  {
    isDisabled: false,
    isCompleted: false,
    component: <StepOne />,
  },
  {
    isDisabled: true,
    isCompleted: false,
    component: <StepTwo />,
  },
  {
    isDisabled: true,
    isCompleted: false,
    component: <StepThree />,
  },
];

const StepWizardComponent = () => <StepWizard steps={steps} />;

generateSnapshots(StepWizardComponent);
