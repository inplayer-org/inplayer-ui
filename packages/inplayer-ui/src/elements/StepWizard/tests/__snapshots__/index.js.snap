// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepWizardComponent renders correctly 1`] = `
.c0 {
  margin-bottom: 2rem;
  text-align: center;
}

.c1 {
  cursor: pointer;
  font-size: 3.125rem;
  line-height: 1;
  margin: 0 1.5rem;
  opacity: 0.4;
  text-shadow: none;
  -webkit-transition: opacity 1s ease,text-shadow 1s ease;
  transition: opacity 1s ease,text-shadow 1s ease;
  will-change: opacity,text-shadow;
}

.c2 {
  cursor: pointer;
  font-size: 3.125rem;
  line-height: 1;
  margin: 0 1.5rem;
  opacity: 0.4;
  text-shadow: none;
  -webkit-transition: opacity 1s ease,text-shadow 1s ease;
  transition: opacity 1s ease,text-shadow 1s ease;
  will-change: opacity,text-shadow;
  pointer-events: none;
}

<div
  className={null}
  style={Object {}}
>
  <div
    className="c0"
  >
    <span
      className="c1"
      onClick={[Function]}
    >
      •
    </span>
    <span
      className="c2"
      onClick={[Function]}
    >
      •
    </span>
    <span
      className="c2"
      onClick={[Function]}
    >
      •
    </span>
  </div>
  <div />
</div>
`;
